### PR TITLE
New main flow  adva

### PR DESCRIPTION
**Pull request by AtomicManic**
Main flow-
1. Still mock IP
2. still old insights and counters
3. refactored "runTest" to smaller functions
I wanted to do some more, but it could wait for the next iteration, i just have to go now. But the flow is fully working.

### CODE REVIEW

File: controller/user.controller.js [hunk 1]
```diff
@@ -43,14 +42,11 @@ const insertExperiment = async (uuid, experiment) => {
   return updatedUser;
 };
 
-const getUserExperiment = (user, experimentId) => {
-  if (!user[0].experiments) {
+const getUserExperiment = (user) => {
```
Comment (peerfikhman) [inline controller/user.controller.js:45]: what is the reson for passin an aray of users and not a single user if anyway we treat only the user[0]

File: newrelic_agent.log [hunk 2]
```diff
@@ -855,3 +855,1243 @@
 {"v":0,"level":30,"name":"newrelic","hostname":"Peer","pid":9744,"time":"2023-02-19T07:50:17.948Z","msg":"Valid event_harvest_config received. Updating harvest cycles. {\"report_period_ms\":5000,\"harvest_limits\":{\"error_event_data\":8,\"log_event_data\":833,\"analytic_event_data\":833,\"custom_event_data\":250}}"}
 {"v":0,"level":30,"name":"newrelic","hostname":"Peer","pid":9744,"time":"2023-02-19T07:50:17.951Z","msg":"Agent state changed from connected to started."}
 {"v":0,"level":30,"name":"newrelic","hostname":"Peer","pid":9744,"time":"2023-02-19T07:50:18.962Z","msg":"Starting initial 1000ms harvest."}
+{"v":0,"level":30,"name":"newrelic","hostname":"TABLET-NDVRD7BI","pid":2828,"time":"2023-02-19T18:56:41.384Z","msg":"Using New Relic for Node.js. Agent version: 9.7.5; Node version: v16.17.1."}
+{"v":0,"level":30,"name":"newrelic","hostname":"TABLET-NDVRD7BI","pid":2828,"time":"2023-02-19T18:56:41.505Z","msg":"Using LegacyContextManager"}
```
Comment (peerfikhman) [inline newrelic_agent.log:859]: This file should be added to gitignore and removed from repository

**Review by peerfikhman — APPROVED** [review]

File: controller/external.controller.js [hunk 3]
```diff
@@ -15,39 +15,65 @@ const {
 } = require("./../Service/route.logic.experiment");
 
 const runTest = async (req, res, next) => {
-  bodyValidator(req);
-  if (!req.body.experimentId) throw new MissingPropertyError("experiment id");
-  if (!(await checkIfExperimentIsActive(req.body.experimentId)))
-    throw new ExperimentNotActive(req.body.experimentId);
+  validateRun(req, req.body.experimentId, req.body.subscription);
```
[the remark this answers is not in the document]
Comment (AtomicManic) [reply controller/external.controller.js:18]: You are correct, everything there is inside that function, maybe i didn't commit it

File: controller/external.controller.js [hunk 4]
```diff
@@ -15,39 +15,65 @@ const {
...
+  if (!experimentId) throw new MissingPropertyError("experiment id");
+  if (!subscription) throw new MissingPropertyError("subscription");
+  if (!(await checkIfExperimentIsActive(experimentId)))
+    throw new ExperimentNotActive(experimentId);
+};
+
+const experimentNewUser = async (req, res, next, experimentId) => {
+  const experiment = await ExperimentRepository.retrieve(experimentId);
   if (!experiment) throw new EntityNotFound("experiment");
   if (checkAttributes(req, experiment, next)) {
```
[the remark this answers is not in the document]
Comment (AtomicManic) [reply controller/external.controller.js:42]: Yes, you are correct. I did talk about it with Suf, who is incharge of that function, to make sure it operates the same response wise

File: controller/external.controller.js [hunk 5]
```diff
@@ -15,39 +15,65 @@ const {
...
+const experimentExistingUser = async (
+  req,
+  res,
+  user,
+  experimentId,
+  subscription
+) => {
+  const experimentsList = getUserExperiment(user);
+  if (!experimentsList) {
+    const variant = await doExperiment(experimentId, user[0].uuid, req);
```
[the remark this answers is not in the document]
Comment (AtomicManic) [reply controller/external.controller.js:61]: It's just the way the user returns from the function. I can change this

File: controller/user.controller.js (hunk 1 above)
Comment (AtomicManic) [reply controller/user.controller.js:45]: You are right, see above

**Pull request merged**